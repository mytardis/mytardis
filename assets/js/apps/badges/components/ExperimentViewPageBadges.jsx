import React, { Fragment, useState } from 'react';
import { createPortal } from 'react-dom';
import PropTypes from 'prop-types';
import { css } from '@emotion/core';
import ExperimentLastUpdatedBadge from './ExperimentLastUpdateBadge';
import PublicAccessBadge from './PublicAccessBadge';
import DatasetCountBadge from './DatasetCountBadge';
import DatafileCountBadge from './DatafileCountBadge';
import { fetchExperimentData } from './utils/FetchData';
import Spinner from './utils/Spinner';
import ExperimentSizeBadge from './ExperimentSizeBadge';


const ExperimentViewPageBadges = ({ experimentID, container, licenseUpdatedCount }) => {
  const [expData, setExpData] = useState('');
  const [isLoading, setIsLoading] = useState(true);
  const spinnerCss = css`
    width: 20%;
    color: 9B9B9B;
    float: right;
  `;

  React.useEffect(() => {
    fetchExperimentData(experimentID).then((data) => {
      setExpData(data);
      setIsLoading(false);
    });
  }, [experimentID, licenseUpdatedCount]);

  return (
    <Fragment>
      {createPortal(
        isLoading ? (
          <Fragment>
            <Spinner override={spinnerCss} />
          </Fragment>
        )
          : (
            <Fragment>
              <span className="mr-2">
                <DatasetCountBadge experimentData={expData} />
              </span>
              <span className="mr-2">
                <DatafileCountBadge experimentData={expData} />
              </span>
              <span className="mr-2">
                <ExperimentSizeBadge experimentData={expData} />
              </span>
              <span className="mr-2">
                <ExperimentLastUpdatedBadge experimentData={expData} />
              </span>
              <span className="mr-2">
                <PublicAccessBadge experimentData={expData} />
              </span>
            </Fragment>
          ), container,
      )
    }
    </Fragment>
  );
};

ExperimentViewPageBadges.propTypes = {
  experimentID: PropTypes.string.isRequired,
  licenseUpdatedCount: PropTypes.number.isRequired,
  container: PropTypes.object.isRequired,
};


export default ExperimentViewPageBadges;
