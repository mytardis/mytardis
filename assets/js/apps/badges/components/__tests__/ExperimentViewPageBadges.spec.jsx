import React from 'react';
import ReactDOM from 'react-dom';
import { act } from '@testing-library/react';
import ExperimentViewPageBadges from '../ExperimentViewPageBadges';

global.fetch = require('jest-fetch-mock');

const fakeExperimentData = {
  update_time: '2020-01-13T14:00:08.908600',
  dataset_count: 1,
  datafile_count: 100,
  public_access: 1,
  experiment_size: 1024,
};
let container = null;
beforeEach(() => {
  jest.spyOn(global, 'fetch').mockImplementation(() => Promise.resolve({
    json: () => Promise.resolve(fakeExperimentData),
  }));
  container = document.createElement('div');
  container.setAttribute('id', 'experiment-1234');
  container.setAttribute('class', 'badges');
  document.body.appendChild(container);
});
afterEach(() => {
  // cleanup on exiting
  container.remove();
  container = null;
});

describe('renders badges on experiment view page', () => {
  it('should render all badges', async () => {
    await act(async () => {
      ReactDOM.render(<ExperimentViewPageBadges
        experimentID="123"
        container={container}
        licenseUpdatedCount={0}
      />, container);
    });
    expect(container).toMatchSnapshot();
    expect(container.querySelectorAll('span').length)
      .toEqual(10);
  });
  it('should render dataset count badge', async () => {
    await act(async () => {
      ReactDOM.render(<ExperimentViewPageBadges
        experimentID="123"
        container={container}
        licenseUpdatedCount={0}
      />, container);
    });
    expect(container.querySelectorAll('span')[0].textContent.trim())
      .toEqual('1');
  });
  it('should render datafile count badge', async () => {
    await act(async () => {
      ReactDOM.render(<ExperimentViewPageBadges
        experimentID="123"
        container={container}
        licenseUpdatedCount={0}
      />, container);
    });
    expect(container.querySelectorAll('span')[2].textContent.trim())
      .toEqual('100');
  });
  it('should render last updated badge', async () => {
    await act(async () => {
      ReactDOM.render(<ExperimentViewPageBadges
        experimentID="123"
        container={container}
        licenseUpdatedCount={0}
      />, container);
    });
    expect(container.querySelectorAll('span')[6].textContent.trim())
      .toEqual('13th January 2020');
  });
  it('should render public access badge', async () => {
    await act(async () => {
      ReactDOM.render(<ExperimentViewPageBadges
        experimentID="123"
        container={container}
        licenseUpdatedCount={0}
      />, container);
    });
    expect(container.querySelectorAll('span')[8].textContent.trim())
      .toEqual('Private');
  });
});
