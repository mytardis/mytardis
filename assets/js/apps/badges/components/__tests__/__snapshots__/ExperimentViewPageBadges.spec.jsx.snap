// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders badges on experiment view page should render all badges 1`] = `
<div
  class="badges"
  id="experiment-1234"
>
  <span
    class="mr-2"
  >
    <span
      class="badge badge-info"
      title="1 dataset"
    >
      <i
        class="fa fa-folder"
      />
       
      1
    </span>
  </span>
  <span
    class="mr-2"
  >
    <span
      class="badge badge-info"
      title="100 files"
    >
      <i
        class="fa fa-file"
      />
       
      100
    </span>
  </span>
  <span
    class="mr-2"
  >
    <span
      class="badge badge-info"
      title="Experiment size is ~ 1 KB"
    >
      1 KB
    </span>
  </span>
  <span
    class="mr-2"
  >
    <span
      class="badge badge-info"
      title="Last updated: Mon, Jan 13, 2020 2:00 PM"
    >
      <i
        class="fa fa-clock-o"
      />
       
      13th January 2020
    </span>
  </span>
  <span
    class="mr-2"
  >
    <span
      class="badge badge-secondary"
      content=" No public access"
      title=" No public access"
    >
      <i
        class="fa fa-eye-slash"
      />
       Private
    </span>
  </span>
</div>
`;
